flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusReporter.java

```diff
@@ -0,0 +1,256 @@
+	private static final int    DEFAULT_PORT = 9249;
+
+	private static final Pattern         UNALLOWED_CHAR_PATTERN = Pattern.compile("[^a-zA-Z0-9:_]");
+	private static final CharacterFilter CHARACTER_FILTER       = new CharacterFilter() {
+		@Override
+		public String filterCharacters(String input) {
+			return replaceInvalidChars(input);
+		}
+	};
+
+	private static final char SCOPE_SEPARATOR = '_';
+
+	private PrometheusEndpoint prometheusEndpoint;
+	private Map<String, Collector> collectorsByMetricName = new HashMap<>();
+
+	@VisibleForTesting
+	static String replaceInvalidChars(final String input) {
+		// https://prometheus.io/docs/instrumenting/writing_exporters/
+		// Only [a-zA-Z0-9:_] are valid in metric names, any other characters should be sanitized to an underscore.
+		return UNALLOWED_CHAR_PATTERN.matcher(input).replaceAll("_");
+	}
+
+	@Override
+	public void open(MetricConfig config) {
+		int port = config.getInteger(ARG_PORT, DEFAULT_PORT);
+		log.info("Using port {}.", port);
+		prometheusEndpoint = new PrometheusEndpoint(port);
+		try {
+			prometheusEndpoint.start(NanoHTTPD.SOCKET_READ_TIMEOUT, true);
+		} catch (IOException e) {
+			log.error("Could not start PrometheusEndpoint on port " + port, e);
+		}
+	}
+
+	@Override
+	public void close() {
+		prometheusEndpoint.stop();
+		CollectorRegistry.defaultRegistry.clear();
+	}
+
+	@Override
+	public void notifyOfAddedMetric(final Metric metric, final String metricName, final MetricGroup group) {
+		final String scope = ((FrontMetricGroup<AbstractMetricGroup<?>>) group).getLogicalScope(CHARACTER_FILTER, SCOPE_SEPARATOR);
+		List<String> dimensionKeys = new LinkedList<>();
+		List<String> dimensionValues = new LinkedList<>();
+		for (final Map.Entry<String, String> dimension : group.getAllVariables().entrySet()) {
+			dimensionKeys.add(CHARACTER_FILTER.filterCharacters(CharMatcher.anyOf("<>").trimFrom(dimension.getKey())));
+			dimensionValues.add(CHARACTER_FILTER.filterCharacters(dimension.getValue()));
+		}
+
+		final String validMetricName = scope + SCOPE_SEPARATOR + CHARACTER_FILTER.filterCharacters(metricName);
+		final String metricIdentifier = group.getMetricIdentifier(metricName);
+		final Collector collector;
+		if (metric instanceof Gauge) {
+			collector = createGauge((Gauge) metric, validMetricName, metricIdentifier, dimensionKeys, dimensionValues);
+		} else if (metric instanceof Counter) {
+			collector = createGauge((Counter) metric, validMetricName, metricIdentifier, dimensionKeys, dimensionValues);
+		} else if (metric instanceof Histogram) {
+			collector = createSummary((Histogram) metric, validMetricName, metricIdentifier, dimensionKeys, dimensionValues);
+		} else if (metric instanceof Meter) {
+			collector = createCounter((Meter) metric, validMetricName, metricIdentifier, dimensionKeys, dimensionValues);
+		} else {
+			log.error("Cannot add unknown metric type: {}. This indicates that the metric type is not supported by this reporter.",
```

Comment:
By convention classes that are part of the metric system never log something as errors but only warnings, the reason being that errors are something that interfere with the job being executed.